biorxiv/journal_tracker/map_preprints_to_journals.py

```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Following the Preprint to Published Path
+
+# The goal of this notebook is to map preprint dois to published dois and published dois to Pubmed Central articles.
+
+# In[1]:
+
+
+from habanero import Crossref
+import pandas as pd
+from ratelimit import limits, sleep_and_retry
+import tqdm
+from urllib.error import HTTPError
+
+
+# In[2]:
+
+
+preprints_df = pd.read_csv("../exploratory_data_analysis/output/biorxiv_article_metadata.tsv", sep="\t")
+preprints_df.head()
+
+
+# In[3]:
+
+
+dois = (
+    preprints_df
+    .doi
+    .unique()
+)
+print(len(dois))
+
+
+# In[4]:
+
+
+cf = Crossref(mailto="dnicholson329@gmail.com")
+
+
+# In[5]:
+
+
+TEN_MINUTES = 600
+
+@sleep_and_retry
+@limits(calls=10, period=TEN_MINUTES)
+def call_crossref(doi_ids):
+    responses = []
+    for doi in doi_ids:
+        try:
+            response = cf.works(ids=doi)
+            responses.append(response)
+        except:
+            responses.append({
+                "message":{
+                    "relation":{"none":"none"}, 
+                    "DOI":doi
+                }
+            })
+        
+    return responses
+
+
+# # Map preprint DOIs to Published DOIs
+
+# In[8]:
+
+
+batch_limit = 100
+doi_mapper_records = []
+
+for batch in tqdm.tqdm(range(0, len(dois), batch_limit)):
+    response = call_crossref(dois[batch:batch+batch_limit])
+    doi_mapper_records += [
+        {
+         "preprint_doi":result['message']['DOI'],
+         "published_doi":(
+                 result['message']['relation']['is-preprint-of'][0]['id'] 
+                 if "is-preprint-of" in result['message']['relation'] 
+                 else ""
+             )
+        }
+        for result in response
+    ]
+
+
+# In[9]:
+
+
+published_doi_df = (
+    pd.DataFrame.from_records(doi_mapper_records)
+    .append(mapped_preprints_df)
+)
+published_doi_df.to_csv("output/mapped_published_doi.tsv", sep="\t", index=False)
+
+
+# # Map Journal Titles to DOI
+
+# In[8]:
+
+
+published_doi_df = pd.read_csv("output/mapped_published_doi.tsv", sep="\t")
+
+
+# In[11]:
+
+
+journal_dois = (
+    published_doi_df
+    .published_doi
+    .unique()
+    .tolist()
+)
+
+
+# In[14]:
+
+
+batch_limit = 100
+journal_mapper_records = []
+
+for batch in tqdm.tqdm(range(0, len(journal_dois), batch_limit)):
```

Comment:
Luckily this is the only time I am using this parser. 

>Curious, is there a reason to read in in batches? Memory?

The main reason here is to throttle my api calls and avoid being blocked/yelled at by the crossref developers. Plus if anything happened to my parser I could track progress and restart midway rather than from the beginning.